yarn-project/pxe/src/synchronizer/synchronizer.ts

```diff
@@ -320,4 +334,67 @@ export class Synchronizer {
       notes: Object.fromEntries(this.noteProcessors.map(n => [n.publicKey.toString(), n.status.syncedToBlock])),
     };
   }
+
+  /**
+   * Retry decoding any deferred notes for the specified contract address.
+   * @param contractAddress - the contract address that has just been added
+   */
+  public async reprocessDeferredNotesForContract(contractAddress: AztecAddress) {
+    const deferredNotes = await this.db.getDeferredNotesByContract(contractAddress);
+
+    // group deferred notes by txHash to properly deal with possible duplicates
+    const txHashToDeferredNotes: Map<TxHash, DeferredNoteDao[]> = new Map();
+    for (const note of deferredNotes) {
+      const notesForTx = txHashToDeferredNotes.get(note.txHash) ?? [];
+      notesForTx.push(note);
+      txHashToDeferredNotes.set(note.txHash, notesForTx);
+    }
+
+    // keep track of decoded notes
+    const newNotes: NoteDao[] = [];
+    // now process each txHash
+    for (const deferredNotes of txHashToDeferredNotes.values()) {
+      // to be safe, try each note processor in case the deferred notes are for different accounts.
+      for (const processor of this.noteProcessors) {
+        const decodedNotes = await processor.decodeDeferredNotes(
+          deferredNotes.filter(n => n.publicKey.equals(processor.publicKey)),
+        );
+        newNotes.push(...decodedNotes);
+      }
+    }
+
+    // now drop the deferred notes, and add the decoded notes
+    await this.db.removeDeferredNotesByContract(contractAddress);
+    await this.db.addNotes(newNotes);
+
+    newNotes.forEach(noteDao => {
+      this.log(
+        `Decoded deferred note for contract ${noteDao.contractAddress} at slot ${
+          noteDao.storageSlot
+        } with nullifier ${noteDao.siloedNullifier.toString()}`,
+      );
+    });
+
+    // now group the decoded notes by public key
+    const publicKeyToNotes: Map<PublicKey, NoteDao[]> = new Map();
+    for (const noteDao of newNotes) {
+      const notesForPublicKey = publicKeyToNotes.get(noteDao.publicKey) ?? [];
+      notesForPublicKey.push(noteDao);
+      publicKeyToNotes.set(noteDao.publicKey, notesForPublicKey);
+    }
+
+    // now for each group, look for the nullifiers in the nullifier tree
+    for (const [publicKey, notes] of publicKeyToNotes.entries()) {
+      const nullifiers = notes.map(n => n.siloedNullifier);
+      const relevantNullifiers: Fr[] = [];
+      for (const nullifier of nullifiers) {
+        // NOTE: this leaks information about the nullifiers I'm interested in to the node.
+        const found = await this.node.findLeafIndex('latest', MerkleTreeId.NULLIFIER_TREE, nullifier);
+        if (found) {
+          relevantNullifiers.push(nullifier);
+        }
+      }
+      await this.db.removeNullifiedNotes(relevantNullifiers, publicKey);
```

Comment:
Great catch!

I think it might be worth refactoring the database in the future to store nullifiers separately of notes so that we might run queries like `where note.nullifier not in nullifiers`

Reply:
Interesting. You're suggesting we keep track of all seen nullifiers? That would remove the need for the `findLeafIndex` call, but the state bloat might concern me. Perhaps if the pattern we want to support is only `note.nullifier not in nullifiers`, we could get away with only storing a bloom filter locally, and fall back to `findLeafIndex` if the filter returns true. 

Broadly though, I don't understand the benefit of the ability to run those kinds of queries if we're already pruning nullified notes.